### PR TITLE
CFY-4067 Add --force flag to 'snapshots restore'.

### PR DESCRIPTION
**Pull request by xwombat**

### CODE REVIEW

File: workflows/cloudify_system_workflows/snapshot.py [hunk 1]
```diff
@@ -385,12 +385,61 @@ def _update_es_node(es_node):
             source['main_file_name'] = ''
 
 
+def _assert_clean_elasticsearch(es):
+    exception = NonRecoverableError('Manager is not clean')
```
Comment (idanmo) [inline workflows/cloudify_system_workflows/snapshot.py]: Perhaps create a method for raising this error where necessary?

File: workflows/cloudify_system_workflows/snapshot.py [hunk 2]
```diff
@@ -385,12 +385,61 @@ def _update_es_node(es_node):
...
+        raise exception
+
+    # check storage
+    storage_scan = elasticsearch.helpers.scan(es, index=_STORAGE_INDEX_NAME)
+    storage_scan = _except_types(storage_scan, 'provider_context', 'snapshot')
+    if next(storage_scan, False):
+        raise exception
+
+
+def _check_conflicts(es, restored_data):
```
Comment (idanmo) [inline workflows/cloudify_system_workflows/snapshot.py]: Please write a short docstring about the conflicts semantics

File: workflows/cloudify_system_workflows/snapshot.py [hunk 3]
```diff
@@ -571,13 +622,19 @@ def recreate_deployments_environments():
...
     mappings = {
         '3.3': _restore_snapshot_format_3_3,
         '3.2': _restore_snapshot_format_3_2
     }
 
     config = _DictToAttributes(config)
+
+    es = _create_es_client(config)
+
+    if not force:
```
Comment (idanmo) [inline workflows/cloudify_system_workflows/snapshot.py]: I think it would be a good idea to check whether the manager is clean anyway, and have a warning log regardless of whether force=true was passed or not. The force=true will skip the validation but the warning will still be there.

File: workflows/cloudify_system_workflows/snapshot.py (hunk 1 above)
Comment (simon-bar) [inline workflows/cloudify_system_workflows/snapshot.py]: Do you mean something like:
def raise(msg):
    raise NonRecoverableError(msg)
?

Comment (idanmo) [inline workflows/cloudify_system_workflows/snapshot.py]: Yes - with a better name for the method. and perhaps write a more informative message that it is only allowed to restore a snapshot on a clean manager.

File: workflows/cloudify_system_workflows/snapshot.py [hunk 4]
```diff
@@ -385,12 +385,86 @@ def _update_es_node(es_node):
...
+    """
+    Check if manager ElasticSearch is clean and raise error (or just
+    log warning) if it isn't.
+
+    :param es: ElasticSearch proxy object
+    :param log_warning: instead raising error just log warning
+    """
+
+    def _raise_or_log():
+        message = ('Manager is not clean, snapshot is allowed to be '
```
Comment (idanmo) [inline workflows/cloudify_system_workflows/snapshot.py]: The message should be different for both cases:
1. force=true: "Forcing snapshot restoration on dirty manager"
2. default: "Snapshot restoration on on a dirty manager is not permitted"

Feel free to adjust the messages.

Comment (idanmo) [inline workflows/cloudify_system_workflows/snapshot.py]: Will be taken care of in a separate PR.

Comment (simon-bar) [inline workflows/cloudify_system_workflows/snapshot.py]: @xwombat please take care of this
